core/src/main/java/org/springframework/security/saml/SAMLAuthenticationProvider.java

```diff
@@ -121,7 +121,7 @@ public Authentication authenticate(Authentication authentication) throws Authent
 
         SAMLCredential authenticationCredential = excludeCredential ? null : credential;
         ExpiringUsernameAuthenticationToken result = new ExpiringUsernameAuthenticationToken(expiration, principal, authenticationCredential, entitlements);
-        result.setDetails(userDetails);
+        result.setDetails(authentication.getDetails());
```

Comment:
Sadly, I don't think that we can main this change inside of a maintenance release since it's possible that there are applications that are calling `getDetails` and expecting the `userDetails` object to come back. By changing it to the result of `AuthenticationDetailsSource`, we'd break those applications.

Instead, what an application would probably need to do is create a custom authentication provider like so:

```java
public class MyAuthenticationProvider implements AuthenticationProvider {
    private final SAMLAuthenticationProvider delegate;

    // .. constructor

    @Override
    public Authentication authenticate(Authentication authentication) {
        ExpiringUsernameAuthenticationToken result = (ExpiringUsernameAuthenticationToken) 
                this.delegate.authenticate(authentication);
        result.setDetails(authentication.getDetails());
        return result;
    }
}
```